pkg/tpkg/registry.go

```diff
@@ -337,97 +349,105 @@ func (gr *gitRegistry) Describe() string {
 	return fmt.Sprintf("%s: %s", gr.name, gr.url)
 }
 
-func (gr *gitRegistry) Load(ctx context.Context, sync bool, cache Cache, ui UI) error {
-	if sync {
-		p := gr.path
-		if gr.path == "" {
-			p = cache.PreferredRegistryPath(gr.url)
-		}
+func (gr *gitRegistry) withFileLock(ctx context.Context, cache Cache, f func(path string) error) error {
+	p := gr.path
+	if gr.path == "" {
+		p = cache.PreferredRegistryPath(gr.url)
+	}
 
-		// Make sure only one pkg-manager is loading the registry at the same time.
-		// Use a lock file in the directory above the registry's checkout path.
-		// This way we don't interfere with cloning/pulling, but still have relatively
-		// good granularity, allowing to sync multiple registries at the same time.
-		lockP := filepath.Join(filepath.Dir(p), ".tpgk_sync.lock")
-		err := os.MkdirAll(filepath.Dir(lockP), 0755)
-		if err != nil {
-			return err
-		}
-		m, err := filemutex.New(lockP)
-		if err != nil {
-			return err
-		}
+	// Make sure only one pkg-manager is loading the registry at the same time.
+	// Use a lock file in the directory above the registry's checkout path.
+	// This way we don't interfere with cloning/pulling, but still have relatively
+	// good granularity, allowing to sync multiple registries at the same time.
+	lockP := filepath.Join(filepath.Dir(p), ".tpgk_sync.lock")
+	err := os.MkdirAll(filepath.Dir(lockP), 0755)
+	if err != nil {
+		return err
+	}
+	m, err := filemutex.New(lockP)
+	if err != nil {
+		return err
+	}
 
-		unlocked := make(chan struct{})
-		ctx, cancel := context.WithTimeout(ctx, time.Minute*3)
-		defer cancel()
-
-		// The following has a race condition:
-		// We could get the lock, then enter the `default` select, but before
-		// closing the channel, the ctx is done and the second select becomes
-		// non-deterministic.
-		// In that case we don't even unlock anymore.
-		// It's a bad case, but better than not giving any error-message.
-		go func() {
-			m.Lock()
-			select {
-			case <-ctx.Done():
-				m.Unlock()
-			default:
-				close(unlocked)
-			}
-		}()
+	unlocked := make(chan struct{})
+	ctx, cancel := context.WithTimeout(ctx, time.Minute*3)
+	defer cancel()
+
+	// The following has a race condition:
+	// We could get the lock, then enter the `default` select, but before
+	// closing the channel, the ctx is done and the second select becomes
+	// non-deterministic.
+	// In that case we don't even unlock anymore.
+	// It's a bad case, but better than not giving any error-message.
+	go func() {
+		m.Lock()
 		select {
-		case <-unlocked:
-			defer m.Unlock()
 		case <-ctx.Done():
-			return fmt.Errorf("unable to acquire sync lock %s", lockP)
+			m.Unlock()
+		default:
+			close(unlocked)
 		}
+	}()
+	select {
+	case <-unlocked:
+		defer m.Unlock()
+	case <-ctx.Done():
+		return fmt.Errorf("unable to acquire sync lock %s", lockP)
+	}
 
-		// Check again whether the directory exists now.
```

Comment:
Did you just drop this comment? It would go in the anonymous function expression in Load, but maybe it is obsolete?

Reply:
I felt it didn't add much, and with the split code it wasn't completely clear where the "again" came from.
It feels natural to test for the existence once you have the lock.